File: common_list/leetcode79.java
package common_list;

import java.util.Scanner;

public class leetcode79{
    public static void main(String[] args){
        Scanner in = new Scanner(System.in);
        char[][] words = {{'A', 'B', 'C', 'E'},
                          {'S', 'F', 'C', 'S'},
                          {'A', 'D', 'E', 'E'}};
        String word = in.nextLine();

        System.out.println(exist(words, word));

        in.close();
    }

    static boolean exist(char[][] board, String word) {
        if(board.length == 0 && board[0].length == 0)
            return false;
        int lenx = board.length, leny = board[0].length;
        boolean[][] judge = new boolean[lenx][leny];
        for(int i = 0; i < lenx; i++){
            for(int j = 0; j < leny; j++){
                judge[i][j] = true;
            }
        }
        for(int x = 0; x < lenx; x++){
            for(int y = 0; y < leny; y++){
                if(board[x][y] == word.charAt(0)){
                    judge[x][y] = false;
                    if(my_exist(board, word, x, y, 0, judge))
                        return true;
                    judge[x][y] = true;
                }
            }
        }
        return false;
    }

    static boolean my_exist(char[][] board, String word, int x, int y, int n, boolean[][] judge){
        if(n == word.length() - 1){
            return true;
        }
        if(x + 1 < board.length && judge[x + 1][y] && board[x + 1][y] == word.charAt(n + 1)){
            judge[x + 1][y] = false;
            if(my_exist(board, word, x + 1, y, n + 1, judge))
                return true;
            judge[x + 1][y] = true;
        }

        if(y + 1 < board[0].length && judge[x][y + 1] && board[x][y + 1] == word.charAt(n + 1)){
            judge[x][y + 1] = false;
            if(my_exist(board, word, x, y + 1, n + 1, judge))
                return true;
            judge[x][y + 1] = true;
        }

        if(x > 0 && judge[x - 1][y] && board[x - 1][y] == word.charAt(n + 1)){
            judge[x - 1][y] = false;
            if(my_exist(board, word, x - 1, y, n + 1, judge))
                return true;
            judge[x - 1][y] = true;
        }

        if(y > 0 && judge[x][y - 1] && board[x][y - 1] == word.charAt(n + 1)){
            judge[x][y - 1] = false;
            if(my_exist(board, word, x, y - 1, n + 1, judge))
                return true;
            judge[x][y - 1] = true;
        }

        return false;
    }
}

File: common_list/leetcode79.go
package common_list

// 给定一个 m x n 二维字符网格 board 和一个字符串单词 word 。如果 word 存在于网格中，返回 true ；否则，返回 false 。

// 单词必须按照字母顺序，通过相邻的单元格内的字母构成，其中“相邻”单元格是那些水平相邻或垂直相邻的单元格。同一个单元格内的字母不允许被重复使用。

// 来源：力扣（LeetCode）
// 链接：https://leetcode-cn.com/problems/word-search
// 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。

// 思路
// 从网格中所有字符串的第一个字符开始进行深度优先搜索
func ExistStr(board [][]byte, word string) bool {
	if len(word) <= 0 {
		return true
	}
	h := len(board)
	if h <= 0 {
		return false
	}
	l := len(board[0])
	found := make([][]bool, h)
	for i := 0; i < h; i++ {
		found[i] = make([]bool, l)
	}

	for i := 0; i < h; i++ {
		for j := 0; j < l; j++ {
			if dfs(board, word, found, i, j) {
				return true
			}
		}
	}

	return false
}

func dfs(board [][]byte, word string, found [][]bool, i, j int) bool {
	if len(word) <= 0 {
		return true
	} else if len(word) == 1 {
		return board[i][j] == word[0]
	}

	if board[i][j] != word[0] {
		return false
	}
	found[i][j] = true
	if i > 0 && !found[i-1][j] {
		if dfs(board, word[1:], found, i-1, j) {
			return true
		}
	}
	if j > 0 && !found[i][j-1] {
		if dfs(board, word[1:], found, i, j-1) {
			return true
		}
	}
	if i < len(board)-1 && !found[i+1][j] {
		if dfs(board, word[1:], found, i+1, j) {
			return true
		}
	}
	if j < len(board[0])-1 && !found[i][j+1] {
		if dfs(board, word[1:], found, i, j+1) {
			return true
		}
	}
	found[i][j] = false
	return false
}
